include/algo/kdtree.h

```diff
@@ -1,20 +1,174 @@
+     * @brief A k-d tree for efficient nearest neighbor search in k-dimensional space.
+     *
+     */
+    template<typename PointT>
+    class KdTree {
+    public:
+        /**
+         * @brief Construct an empty k-d tree.
+         */
+        KdTree(): root_(nullptr), dim_(PointT::DIM) {}
+
+        /**
+         * @brief Construct and build the k-d tree from a point cloud.
+         * @param points Vector of input points.
+         * @param dim Dimensionality of the space. If -1, defaults to 3.
+         */
+        KdTree(const std::vector<PointT>& points, int dim = -1): root_(nullptr) {
+            build(points, dim);
+        }
+
+        /**
+         * @brief Destructor.
+         */
+        ~KdTree() {
+            clear();
+        }
+
+        /**
+         * @brief Build the k-d tree from a given point set.
+         * @param points Vector of input points.
+         * @param dim Dimensionality of the space. If -1, defaults to 3.
+         */
+        void build(const std::vector<PointT>& points, int dim = -1) {
+            clear();
+            points_ = points;
+
+            dim_ = (dim > 0) ? dim : 3;
+
+            std::vector<int> indices(points.size());
+            std::iota(std::begin(indices), std::end(indices), 0);
+
+            root_ = buildRecursive(indices.data(), static_cast<int>(points.size()), 0);
+        }
+
+        /**
+         * @brief Clear the tree.
+         */
+        void clear() {
+            clearRecursive(root_);
+            root_ = nullptr;
+            points_.clear();
+        }
+
+        /**
+         * @brief Internal node structure of the k-d tree.
+         */
+        struct Node {
+            int idx;        // index to the original point
+            Node* next[2];  // pointers to the child nodes
+            int axis;       // dimension's axis
+
+            Node(): idx(-1), axis(-1) {
+                next[0] = next[1] = nullptr;
+            }
+        };
+
+        /**
+         * @brief Internal exception class for tree errors.
+         */
+        class Exception : public std::exception {
+            using std::exception::exception;
+        };
+
+        /**
+         * @brief Search for the nearest neighbor to a query point.
+         *
+         * @param query The query point.
+         * @param minDist Optional output parameter for the squared distance to the nearest point.
+         * @return Index of the nearest neighbor in the original input point vector.
+         */
+        int search(const PointT& query, double* minDist = nullptr) const {
+            int guess;
+            double _minDist = std::numeric_limits<double>::max();
+
+            searchRecursive(query, root_, &guess, &_minDist);
+
+            if (minDist) *minDist = _minDist;
+            return guess;
+        }
+
+    private:
+        /**
+         * @brief Recursively build the tree from a set of point indices.
+         */
+        Node* buildRecursive(int* indices, int npoints, int depth) {
```

Comment:
I imagine there should be a better way to pass `indices` here. Surely we can use `std::vector`'s `begin` and `end`?